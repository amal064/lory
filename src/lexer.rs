use std::{iter, str::Chars};

use phf::phf_map;
use thiserror::Error;

use crate::{ast::TokenKind, span::Span};

#[derive(Debug)]
#[non_exhaustive]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

pub struct Lexer<'src> {
    src: &'src str,
    chars: Chars<'src>,
}

#[derive(Debug, Error)]
pub enum Error {
    #[error("string not terminated")]
    UnterminatedString { pos: usize },
    #[error("unexpected character")]
    UnexpectedCharacter,
}

impl<'src> Lexer<'src> {
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            src: source,
            chars: source.chars(),
        }
    }

    #[inline]
    fn advance(&mut self) -> Option<char> {
        self.chars.next()
    }

    #[inline]
    fn first(&self) -> Option<char> {
        self.chars.clone().next()
    }

    #[inline]
    fn second(&self) -> Option<char> {
        let mut iter = self.chars.clone();
        iter.next();
        iter.next()
    }

    #[inline]
    fn pos(&self) -> usize {
        self.src.len() - self.chars.as_str().len()
    }

    #[allow(clippy::too_many_lines)]
    fn advance_token(&mut self) -> Result<Token, Error> {
        loop {
            let start = self.pos();
            let Some(first_char) = self.advance() else {
                return Ok(Token {
                    kind: TokenKind::Eof,
                    span: Span::new(0, 0),
                });
            };
            let token_kind = match first_char {
                c if c.is_whitespace() => continue,
                '(' => TokenKind::LeftParen,
                ')' => TokenKind::RightParen,
                '{' => TokenKind::LeftBrace,
                '}' => TokenKind::RightBrace,
                '[' => TokenKind::LeftBracket,
                ']' => TokenKind::RightBracket,
                ',' => TokenKind::Comma,
                '.' => TokenKind::Dot,
                '-' => TokenKind::Minus,
                '+' => TokenKind::Plus,
                ';' => TokenKind::Semicolon,
                '*' => TokenKind::Star,
                '!' => match self.first() {
                    Some('=') => {
                        self.advance();
                        TokenKind::BangEqual
                    }
                    _ => TokenKind::Bang,
                },
                '=' => match self.first() {
                    Some('=') => {
                        self.advance();
                        TokenKind::EqualEqual
                    }
                    _ => TokenKind::Equal,
                },
                '<' => match self.first() {
                    Some('=') => {
                        self.advance();
                        TokenKind::LessEqual
                    }
                    _ => TokenKind::Less,
                },
                '>' => match self.first() {
                    Some('=') => {
                        self.advance();
                        TokenKind::GreaterEqual
                    }
                    _ => TokenKind::Greater,
                },
                '/' => match self.first() {
                    Some('/') => {
                        while let Some(c) = self.advance() {
                            if c == '\n' {
                                break;
                            }
                        }
                        continue;
                    }
                    _ => TokenKind::Slash,
                },
                '"' => {
                    let mut terminated = false;
                    while let Some(c) = self.advance() {
                        if c == '"' {
                            terminated = true;
                            break;
                        }
                    }
                    if !terminated {
                        return Err(Error::UnterminatedString { pos: start });
                    }
                    TokenKind::String {
                        string: self.src[start + 1..self.pos() - 1].to_string(),
                    }
                }
                c if c.is_ascii_digit() => {
                    let mut has_dot = false;
                    loop {
                        match self.first() {
                            Some('.') => {
                                if has_dot {
                                    break;
                                }
                                has_dot = true;

                                match self.second() {
                                    Some(c) if c.is_ascii_digit() => {}
                                    _ => break,
                                }
                            }
                            Some(c) if c.is_ascii_digit() => {}
                            _ => break,
                        };
                        self.advance();
                    }
                    let num = self.src[start..self.pos()].parse().unwrap();
                    TokenKind::Number(num)
                }
                c if c.is_ascii_alphabetic() || c == '_' => {
                    while let Some(c) = self.first() {
                        if !c.is_ascii_alphanumeric() && c != '_' {
                            break;
                        }
                        self.advance();
                    }

                    let ident =
                        self.src[start..self.src.len() - self.chars.as_str().len()].to_string();
                    KEYWORDS
                        .get(&ident[..])
                        .cloned()
                        .unwrap_or(TokenKind::Identifier { ident })
                }
                _ => return Err(Error::UnexpectedCharacter),
            };
            let end = self.src.len() - self.chars.as_str().len();
            return Ok(Token {
                kind: token_kind,
                span: Span::new(start, end),
            });
        }
    }
}

pub fn tokenize(input: &str) -> impl Iterator<Item = Result<Token, Error>> + '_ {
    let mut lexer = Lexer::new(input);
    iter::from_fn(move || {
        let token = lexer.advance_token();
        if let Ok(Token {
            kind: TokenKind::Eof,
            ..
        }) = token
        {
            None
        } else {
            Some(token)
        }
    })
}

static KEYWORDS: phf::Map<&'static str, TokenKind> = phf_map! {
    "and" => TokenKind::And,
    "class" => TokenKind::Class,
    "else" => TokenKind::Else,
    "false" => TokenKind::False,
    "for" => TokenKind::For,
    "fun" => TokenKind::Fun,
    "if" => TokenKind::If,
    "nil" => TokenKind::Nil,
    "or" => TokenKind::Or,
    "print" => TokenKind::Print,
    "return" => TokenKind::Return,
    "super" => TokenKind::Super,
    "this" => TokenKind::This,
    "true" => TokenKind::True,
    "var" => TokenKind::Var,
    "while" => TokenKind::While,
};
