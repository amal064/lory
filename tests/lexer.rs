use insta::assert_debug_snapshot;
use lory::lexer;

#[test]
fn identifiers() {
    let tokens = lexer::tokenize(include_str!("lexer/identifiers.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn keywords() {
    let tokens = lexer::tokenize(include_str!("lexer/keywords.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn numbers() {
    let tokens = lexer::tokenize(include_str!("lexer/numbers.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn punctuators() {
    let tokens = lexer::tokenize(include_str!("lexer/punctuators.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn strings() {
    let tokens = lexer::tokenize(include_str!("lexer/strings.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn whitespace() {
    let tokens = lexer::tokenize(include_str!("lexer/whitespace.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn comments() {
    let tokens = lexer::tokenize(include_str!("lexer/comments.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}

#[test]
fn unexpectedchar() {
    let tokens = lexer::tokenize(include_str!("lexer/unexpectedchar.lox"));
    assert_debug_snapshot!(tokens.collect::<Vec<_>>());
}
