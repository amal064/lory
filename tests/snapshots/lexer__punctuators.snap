---
source: tests/lexer.rs
expression: "tokens.collect::<Vec<_>>()"
---
[
    Ok(
        Token {
            kind: LeftParen,
            span: Span {
                start: 0,
                end: 1,
            },
        },
    ),
    Ok(
        Token {
            kind: RightParen,
            span: Span {
                start: 1,
                end: 2,
            },
        },
    ),
    Ok(
        Token {
            kind: LeftBrace,
            span: Span {
                start: 2,
                end: 3,
            },
        },
    ),
    Ok(
        Token {
            kind: RightBrace,
            span: Span {
                start: 3,
                end: 4,
            },
        },
    ),
    Ok(
        Token {
            kind: Semicolon,
            span: Span {
                start: 4,
                end: 5,
            },
        },
    ),
    Ok(
        Token {
            kind: Comma,
            span: Span {
                start: 5,
                end: 6,
            },
        },
    ),
    Ok(
        Token {
            kind: Plus,
            span: Span {
                start: 6,
                end: 7,
            },
        },
    ),
    Ok(
        Token {
            kind: Minus,
            span: Span {
                start: 7,
                end: 8,
            },
        },
    ),
    Ok(
        Token {
            kind: Star,
            span: Span {
                start: 8,
                end: 9,
            },
        },
    ),
    Ok(
        Token {
            kind: BangEqual,
            span: Span {
                start: 9,
                end: 11,
            },
        },
    ),
    Ok(
        Token {
            kind: EqualEqual,
            span: Span {
                start: 11,
                end: 13,
            },
        },
    ),
    Ok(
        Token {
            kind: LessEqual,
            span: Span {
                start: 13,
                end: 15,
            },
        },
    ),
    Ok(
        Token {
            kind: GreaterEqual,
            span: Span {
                start: 15,
                end: 17,
            },
        },
    ),
    Ok(
        Token {
            kind: BangEqual,
            span: Span {
                start: 17,
                end: 19,
            },
        },
    ),
    Ok(
        Token {
            kind: Less,
            span: Span {
                start: 19,
                end: 20,
            },
        },
    ),
    Ok(
        Token {
            kind: Greater,
            span: Span {
                start: 20,
                end: 21,
            },
        },
    ),
    Ok(
        Token {
            kind: Slash,
            span: Span {
                start: 21,
                end: 22,
            },
        },
    ),
    Ok(
        Token {
            kind: Dot,
            span: Span {
                start: 22,
                end: 23,
            },
        },
    ),
]
